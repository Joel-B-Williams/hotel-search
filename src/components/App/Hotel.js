import React, { Component } from 'react';

class Hotel extends Component {

  defaultImage = (ev) => {
    ev.target.src = "https://i.imgur.com/qUvaEAl.png"
  }

  render () {
    return (
      <div className="hotel-list">
        <img src={this.props.hotelStaticContent.mainImage.url} onError={this.defaultImage} className="photo"/>
        <div className="details">
          <div className="hotel-name">
            {this.props.hotelStaticContent.name || "No name found" }
          </div>
          <div className="hotel-location">
            {this.props.hotelStaticContent.neighborhoodName || "Location unknown"}
          </div>
        </div>
        <div className="price">
          <table className="price-table">
            <tbody>
              <tr>
                <td>Per night: </td>
                <td className="price-per-night">
                  {/* find less risky method if possible */}
                  <p dangerouslySetInnerHTML={{__html: this.props.lowestAveragePrice.symbol}}></p>
                  {this.props.lowestAveragePrice.amount || "$$"}
                </td>
              </tr>
              <tr className="points-row">
                <td>Points earned:</td>
                <td className="points-earned">
                  {this.props.rewards.miles || "Reward unavailable"}
                </td>
              </tr>
            </tbody>
          </table>
          <button className="btn hotel-btn">Select Now</button>
        </div>
      </div>
    );
  }
}
  
export default Hotel;