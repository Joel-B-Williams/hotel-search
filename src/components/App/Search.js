import React, { Component } from 'react';

class Search extends Component {

  onChange  = event => {
    this.props.handleChange(event.target.value)
  }

  render() {
    return(
        <div className="filters">
            <label htmlFor="propertySearch">Search properties</label>
            <input 
              type="text" 
              onKeyUp={this.onChange} 
              id="propertySearch"
              placeholder="Property name"/>
            <button onClick={this.props.handleClick} className="btn">Price Low-to-High</button>
        </div>
    )
  }
}

export default Search;