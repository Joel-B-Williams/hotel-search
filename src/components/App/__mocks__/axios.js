'use strict';
module.exports = {
  get: () => {
    return Promise.resolve({
      data: [
        {"id":"907","rewards":{"miles":10000},"lowestAveragePrice":{"currency":"USD","symbol":"&#36;","amount":579.0},"hotelStaticContent":{"mainImage":{"category":"EXTERIOR","url":"http://d2whcypojkzby.cloudfront.net/imageRepo/2/0/68/56/314/ExteriorCarsGone_S.jpg"},"name":"Omni Chicago Hotel & Suites Magnificent Mile","neighborhoodName":"Magnificent Mile"}},
        {"id":"908","rewards":{"miles":8000},"lowestAveragePrice":{"currency":"USD","symbol":"&#36;","amount":234.0},"hotelStaticContent":{"mainImage":{"category":"EXTERIOR","url":"http://d2whcypojkzby.cloudfront.net/imageRepo/6/0/96/743/239/chidt-exterior-0227-hor-clsc_S.jpg"},"name":"Chicago Marriott Downtown Magnificent Mile","neighborhoodName":"Magnificent Mile"}},{"id":"87390","rewards":{"miles":12000},"lowestAveragePrice":{"currency":"USD","symbol":"&#36;","amount":123.0},"hotelStaticContent":{"mainImage":{"url":"http://d2whcypojkzby.cloudfront.net/imageRepo/1/0/16/99/327/012857_A_P.jpg"},"name":"Homewood Suites by Hilton Chicago-Lincolnshire","neighborhoodName":"Lincolnshire"}}
      ]
    });
  }
};