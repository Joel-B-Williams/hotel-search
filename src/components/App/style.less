html, body {
    margin: 0;
    padding: 0;
    background: #DCDCDC;
}
.app-container {
    font-family: Helvetica, Arial, sans-serif;
    
    .header {
        padding: 20px;
    }
    
    .content {
        display: grid;
        grid-template-columns: 20% 75%;
        grid-column-gap: 20px;
        padding: 20px;
        background: #DCDCDC;
        
        .btn {
            background-color: #6DBA4A;
            color: white;
            border-radius: 5px;
            margin: 4px;
            padding: 4px;
            
            &:hover {
                background-color: #7cc15c;
                cursor: pointer;
            }
        }
        
        .search-area {
            background: white;
            min-height: 95vh;
        }

        .filters {
            position: fixed;
            width: 20%;
            margin: 4px;
            
            label {
                color: gray;
                font-size: 14px;
                margin: 4px;
            }

            input {
                margin: 4px;
            }
            
        }

        .hotel-list {
            background: white;
            display: grid;
            grid-template-columns: 20% 1fr 20%;
            border-bottom: solid 10px #DCDCDC;
            
            .photo {
                width: 100%;
                height: 180px;
            }
            
            .details {
                padding: 10px;
                
                .hotel-name {
                    font-size: 16px;
                    margin: 4px;
                }
                
                .hotel-location {
                    color: gray;
                    font-size: 12px;
                    margin: 4px;
                }
            }            
            .price {
                color: #6DBA4A;
                padding: 10px;
                border-left: solid gray 1px;
            }
            .price-table {
                font-size: 12px;

                p {
                    display: inline-block;
                }
                    
                .points-row {
                    color: #2E49AC;
                }

                .price-per-night, .points-earned {
                    font-weight: bold;
                    font-size: 18px;
                }
            }
            
            .hotel-btn {
                font-size: 18px;
            }
        }
            
        .hotel-list:last-of-type {
            border-bottom: none;
            
        }
    }
}