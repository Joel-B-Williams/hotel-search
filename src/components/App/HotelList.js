import React, { Component } from 'react';
import Hotel from './Hotel'

class HotelList extends Component {

  render() {
    const error = this.props.error;
    const errorMessage = 
    <div className="status-message">
      <p>Something has gone horribly pear-shaped</p>
      <p>Please try again</p>
    </div>
    
    return (
      <div>
        {error ? 
        (errorMessage) : 
        (this.props.hotels
        .filter(hotel => hotel.hotelStaticContent.name.toUpperCase().indexOf(this.props.searchTerm.toUpperCase()) >= 0)
        .map((hotel) => (<Hotel key={hotel.id} {...hotel} />)) )}
      </div>
    )        
  }
}


export default HotelList;