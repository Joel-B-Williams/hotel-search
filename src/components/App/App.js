import React, { Component } from 'react';
import axios from 'axios';
import './style.less';

import HotelList from './HotelList';
import Search from './Search';

class App extends Component {
    state = {
        hotels: [],
        searchTerm: '',
        error: false 
    }

    componentDidMount() {
        axios
        .get('https://homework-app.rocketmiles.com/fe-homework/rates')
        .then(res => this.setState({ hotels: res.data.results.hotels }))
        .catch(err => this.setState({ error: true }));
    }

    handleChange = (searchTerm) => {
        this.setState({ searchTerm: searchTerm })
    }

    handleClick = () => {
        let hotels = this.state.hotels
        hotels.sort((a,b) => a.lowestAveragePrice.amount - b.lowestAveragePrice.amount)
        this.setState({ hotels: hotels })
    }

    render() {
        return (
            <div className="app-container">
                <div className="content">
                    <div className="search-area">
                        <Search 
                            handleChange={this.handleChange} 
                            handleClick={this.handleClick} 
                        />
                    </div>
                    <HotelList 
                        hotels={this.state.hotels}
                        searchTerm={this.state.searchTerm}
                        error={this.state.error}
                    />
                </div>
            </div>
        )
    }
}

export default App;
