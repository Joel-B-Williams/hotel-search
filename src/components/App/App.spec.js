import React from 'react';
import { shallow } from 'enzyme';
import App from './App';
import { create } from 'react-test-renderer';
import axios from 'axios';

jest.mock('axios');

describe('App', () => {
    const wrapper = shallow(<App />);
    
    const list = [
        {"id":"907","rewards":{"miles":10000},"lowestAveragePrice":{"currency":"USD","symbol":"&#36;","amount":579.0},"hotelStaticContent":{"mainImage":{"category":"EXTERIOR","url":"http://d2whcypojkzby.cloudfront.net/imageRepo/2/0/68/56/314/ExteriorCarsGone_S.jpg"},"name":"Omni Chicago Hotel & Suites Magnificent Mile","neighborhoodName":"Magnificent Mile"}},
        {"id":"908","rewards":{"miles":8000},"lowestAveragePrice":{"currency":"USD","symbol":"&#36;","amount":234.0},"hotelStaticContent":{"mainImage":{"category":"EXTERIOR","url":"http://d2whcypojkzby.cloudfront.net/imageRepo/6/0/96/743/239/chidt-exterior-0227-hor-clsc_S.jpg"},"name":"Chicago Marriott Downtown Magnificent Mile","neighborhoodName":"Magnificent Mile"}},{"id":"87390","rewards":{"miles":12000},"lowestAveragePrice":{"currency":"USD","symbol":"&#36;","amount":123.0},"hotelStaticContent":{"mainImage":{"url":"http://d2whcypojkzby.cloudfront.net/imageRepo/1/0/16/99/327/012857_A_P.jpg"},"name":"Homewood Suites by Hilton Chicago-Lincolnshire","neighborhoodName":"Lincolnshire"}}
    ]
    
    it('renders the component', () => {
        expect(wrapper.find('.app-container').exists()).toBe(true);
    });
    
    it('matches snapshot', () => {
        const tree = create(<App />);
        expect(tree.toJSON()).toMatchSnapshot();
    });
    
    it('fetches a list of hotels', () => {
        const getSpy = jest.spyOn(axios, 'get');
        const app = shallow(<App/>);
        
        expect(getSpy).toBeCalled();
    });

    it('sorts by price', () => {

        wrapper.setState({hotels: list});
        const unsortedList = wrapper.state('hotels');

        expect(unsortedList[0].id).toEqual("907");
        
        wrapper.instance().handleClick();
        const sortedList = wrapper.state('hotels');
        
        expect(sortedList[0].id).toEqual("87390");
    });
});