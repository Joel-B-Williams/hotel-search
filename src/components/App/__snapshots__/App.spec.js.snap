// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App matches snapshot 1`] = `
<div
  className="app-container"
>
  <div
    className="content"
  >
    <div
      className="search-area"
    >
      <div
        className="filters"
      >
        <label
          htmlFor="propertySearch"
        >
          Search properties
        </label>
        <input
          id="propertySearch"
          onKeyUp={[Function]}
          placeholder="Property name"
          type="text"
        />
        <button
          className="btn"
          onClick={[Function]}
        >
          Price Low-to-High
        </button>
      </div>
    </div>
    <div>
      <div />
      <div>
        <div
          className="status-message"
        >
          <p>
            No properties found
          </p>
          <p>
            Please adjust search parameters
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
